electrum_nmc/electrum/names.py

```diff
@@ -132,6 +132,296 @@ def validate_value_length(value: bytes):
+        elif re.search(r'(?<!xn)--(?!-)', address):
+            raise ValueError("Sequence '--' is not allowed except in 'xn--'")
+        # Check if it ends with '.b32.i2p'
+        elif not address.endswith('.b32.i2p'):
+            raise ValueError("Address doesn't end with '.b32.i2p'")
+
+        # Check length of the address (According to https://geti2p.net/en/docs/naming "grep 67 characters")
+        if len(address) > 67:
+            raise ValueError("Address length exceeds 67 characters")
+        
+        # remove '.b32.i2p'
+        address = address[:-len('.b32.i2p')]
+
+        # If starts with 'xn--', remove this prefix
+        if address.startswith('xn--'):
+            address = address[4:]
+
+        # Only [a-z] [0-9] '.' and '-' are allowed
+        address_lower = address.lower()
+        pattern = r'^[a-z0-9.-]+$'
+        if not re.match(pattern, address_lower):
+            raise ValueError("Invalid characters in the address")
+
+        # Add padding
+        padding_needed = (8 - len(address) % 8)
+        padded_address = address + '=' * padding_needed
+
+        decoded = base64.b32decode(padded_address, casefold=True)
+        if len(decoded) != BASE32_LENGTH:
+            raise ValueError("Invalid length for base32 hash")
+
+    except Exception as e:
+        raise ValueError(f"Invalid I2P address: {e}")
+
+def validate_relative_domain(value: str, /, *, rfc_1034: bool = False, rfc_2782: bool = False):
+        if not value:
+            return False
+        try:
+            return not re.search(r"\s", value) and re.match(
+                # First character of the domain
+                rf"^(?:[a-zA-Z0-9{'_'if rfc_2782 else ''}]"
+                # Sub domain + hostname
+                + r"(?:[a-zA-Z0-9-_]{0,61}[A-Za-z0-9])?\.)"
+                # First 61 characters of the gTLD
+                + r"+[A-Za-z0-9][A-Za-z0-9-_]{0,61}"
+                # Last character of the gTLD
+                + rf"[A-Za-z]{r'.$' if rfc_1034 else r'$'}",
+                value.encode("idna").decode("utf-8"),
+                re.IGNORECASE,
+            )
+        except UnicodeError:
+            return False
+
+def validate_domains(domain: str) -> None:
+    if domain.endswith('.'):
+        relative_domain = domain[:-1]
+        if not validate_relative_domain(relative_domain):
+            raise ValueError(f'Invalid domain: {domain}')
+
+    else:
+        if not validate_relative_domain(domain):
+            raise ValueError(f'Invalid domain: {domain}')
+
+        try:
+            # Read the TLD set from the JSON file
+            current_directory = os.path.dirname(os.path.abspath(__file__))
+            json_file_path = os.path.join(current_directory, 'tld_list.json')
+
+            with open(json_file_path, 'r') as f:
+                tld_set = set(json.load(f))
+        except Exception:
+            # If the JSON file is not found, we will skip the TLD validation
+            return
```

Comment:
I'd prefer that we not silently fail here; if the JSON file is missing something has gone wrong in the build process.